### DISCUSSION TITLE
Twilio::CallbackController:create not found

### DISCUSSION THREAD

**Discussion by masdeval** [chatwoot/chatwoot#1848] [Q&A]
**Describe the bug**

I have configured a WhatsApp inbox bounded to Twilio. On my server, the error is in Twilio::CallbackController :

```

    -Started POST "/twilio/callback" for 34.235.134.132 at 2021-01-28 18:26:41 -0500
    -Processing by Twilio::CallbackController#create as */*
    - Parameters: {"SmsMessageSid"=>"SMe5540131c50fb879622b0e8d08ead2fb", "NumMedia"=>"0", "SmsSid"=>"SMe5540131c50fb879622b0e8d08ead2fb", "SmsStatus"=>"received", "Body"=>"Hello", "To"=>"whatsapp:+14155238886", "NumSegments"=>"1", "MessageSid"=>"SMe5540131c50fb879622b0e8d08ead2fb", "AccountSid"=>"AC4e285d6b423e9a5c384275957cf4604b", "From"=>"whatsapp:+556781189719", "ApiVersion"=>"2010-04-01"}
    -Can't verify CSRF token authenticity.
    - Completed 404 Not Found in 26ms (Views: 0.2ms | ActiveRecord: 6.9ms | Allocations: 3021)


```
In the Twilio platform the message is:

    `Error 11200: There was a failure attempting to retrieve the contents of this URL.

`
The address of my installation is masdeval.ddns.net:3000.
The callback I am using is http://masdeval.ddns.net:3000/twilio/callback 

It seems there is no webservice answering to the path /twilio/callback.

**To Reproduce**

**Expected behavior**

WhatsApp inbox working in chatwoot using twilio sandbox for WhatsApp

**Screenshots**

**Browser logs**

Share the browser logs to debug the issue further

**Server logs**

Share the server logs to debug the issue further

**Environment**

Its a self hosted installation on Linux 18.04.

**Desktop (please complete the following information):**
 - OS: Linux
 - Browser: Firefox
 - Version: 18.04

**Comment by pranavrajs** [comment]
@masdeval Can you please run `::Channel::TwilioSms.all` in rails console and see if you have Twilio Channel registered correctly. Incoming `account_sid` and `phone_number` has to match, otherwise, a 404 will be returned.

**Comment by masdeval** [comment]
It's returning an empty brackets => []

**Comment by pranavrajs** [comment]
You have to create an inbox as described here https://www.chatwoot.com/docs/channels/whatsapp-sms-twilio

**Comment by masdeval** [comment]
I have done this already. You can have access to my installation at HTTP://masdeval.ddns.net:3000

**Comment by pranavrajs** [comment]
@masdeval deleted the last comment as it has login credential.

**Comment by masdeval** [comment]
Where could I find logging for the inbox creation process? I would like to verify whether there was some error during the channel creation. 
I found the chatwoot/log/development.log file and there is nothing there regarding this.

**Comment by pranavrajs** [comment]
If you are default environment variable, production.log might not be generated. You could see the logs by `tail -f /var/log/syslog`

**Comment by jeovajireh** [comment]
> You have to create an inbox as described here https://www.chatwoot.com/docs/channels/whatsapp-sms-twilio

I also followed this tutorial, everything is created, both in Twilio and chatwoot, but same thing, there is nothing in the rails console,

[7] pry(main)> ::Channel::TwilioSms.all
=> []

 The URL with final /twilio/callback does not exist.


"The page you were looking for doesn't exist.
You may have mistyped the address or the page may have moved.
If you are the application owner check the logs for more information."

Thanks

**Comment by jeovajireh** [comment]
> If you are default environment variable, production.log might not be generated. You could see the logs by `tail -f /var/log/syslog`

Is there a command on the console to create the path /twilio/callback for Twilio? 
The route to Twitter /twitter/callback exists in my installation.

Thanks

**Comment by jeovajireh** [comment]
There is something very strange about the project that the developers have not yet observed or documented. I spent last night installing versions with Docker and testing to see if any of them raised the endpoint for Twilio and nothing, went to the chatwoot / chatwoot version: v1.3.0, after each test I ran the following commands:

docker container rm $ (docker container ls -aq)
docker system prune -a

and

DROP DATABASE chatwoot_production;

In none of the versions was the endpoint /twilio/callback present, all the same error.

"The page you were looking for doesn't exist.
You may have mistyped the address or the page may have moved.
If you are the application owner check the logs for more information."

In the syslog file nothing that indicates an error when creating the Twilio inbox.

The Twitter endpoint is already there, but it does not say in the documentation that the Twilio endpoint is only created if Twilio returns something.

**Comment by jeovajireh** [comment]
After a lot of reinstalling, now I have the following result:

[1] pry (main)>::Channel::TwilioSms.all
=> #<Channel::TwilioSms::ActiveRecord_Relation:0x11d64>


Is this progress?

The only error in /var/log/syslog

Can't verify CSRF token authenticity.

**Comment by jeovajireh** [comment]
One question, does anyone in the world use Twilio in chatwoot or is it just decoration?


https://app.chatwoot.com/twilio/callback

"The page you were looking for doesn't exist.

You may have mistyped the address or the page may have moved.

If you are the application owner check the logs for more information."

And for anyone to think they’re in the wrong place.


Twitter

https://app.chatwoot.com/twitter/callback

**Comment by pranavrajs** [comment]
@viapratica I don't understand what you are trying to accomplish. This URL(https://app.chatwoot.com/twilio/callback) will give you a 404 in the following cases.

1. If you hit URL from the browser, this URL does not accept GET request. It accept only POST request. You have to try it with Postman or any other tool similar to that.
2. If your Twilio Credentials does not match. There needs an exact match of AccountSid and PhoneNumber. If those don't match, it will return a 404. 

From the logs above and the issue, it is not clear what the problem is. We use Twilio in production. 

Can you clarify what the problem exactly is?

1. Can you create a Twilio Inbox?
2. Are you getting messages from Twilio?
3. Can you respond to messages?

**Comment by jeovajireh** [comment]
I don't know where I got lost in previous installations, my Linux is Ubuntu 18.04, the first time I did the installation there was no way out on the rails console:

7] pry(main)> ::Channel::TwilioSms.all
=> []

but in this last installation, there was a way out:

[1] pry (main)>::Channel::TwilioSms.all
=> #Channel::TwilioSms::ActiveRecord_Relation:0x11d64

From yesterday to today I deleted the Twilio integration and added it again and everything worked wonderfully!

Congratulations!

I am Brazilian, if you need any help with the Brazilian Portuguese translation I am available.

Thank you for your attention

**Reply by pomborenato** [reply]
Bom dia, também sou do Brasil e estou com o mesmo problema de a url de callback dar not found, pode me ajudar?

**Comment by jeovajireh** [comment]
Yes, you are right, I had done the tests with GET and not with POST. Excuse me!

**Comment by masdeval** [comment]
Olá.

Não consegui resolver. Uma pessoa no forum comentou que reinstalou tudo e
funcionou.

**Reply by pomborenato** [reply]
Eu usei o deploy do heroku e vc fez como?

**Comment by masdeval** [comment]
Instalei no Ubuntu

**Comment by ciclanoio** [comment]
Alguém conseguiu resolver isso?  alguns anos depois e mesmo problema
